### PR TITLE
gh-79846: Make `ssl.create_default_context()` ignore invalid certificates

### LINKED ISSUE #79846: Function ssl.create_default_context raises exception on Windows 10  when called with  ssl.Purpose.SERVER_AUTH) attribute
BPO | [35665](https://bugs.python.org/issue35665)
--- | :---
Nosy | @tiran, @pervlad, @MDM-79, @pukkandan
Files | <li>[cacerts.pem](https://bugs.python.org/file48029/cacerts.pem "Uploaded as text/plain at 2019-01-07.18:44:26 by @pervlad"): pub certs</li><li>[Untitled.png](https://bugs.python.org/file50311/Untitled.png "Uploaded as image/png at 2021-09-28.15:39:56 by @MDM-79"): MUPCA Root - Certificates</li>

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = None
closed_at = <Date 2019-01-08.09:38:24.479>
created_at = <Date 2019-01-05.12:24:45.346>
labels = ['expert-SSL', 'type-bug', '3.9']
title = 'Function ssl.create_default_context raises exception on Windows 10  when called with  ssl.Purpose.SERVER_AUTH) attribute'
updated_at = <Date 2021-09-28.21:47:58.903>
user = 'https://github.com/pervlad'
```

bugs.python.org fields:
```python
activity = <Date 2021-09-28.21:47:58.903>
actor = 'pukkandan'
assignee = 'none'
closed = True
closed_date = <Date 2019-01-08.09:38:24.479>
closer = 'pervlad'
components = ['SSL']
creation = <Date 2019-01-05.12:24:45.346>
creator = 'pervlad'
dependencies = []
files = ['48029', '50311']
hgrepos = []
issue_num = 35665
keywords = []
message_count = 16.0
messages = ['333054', '333055', '333158', '333159', '333168', '333181', '333183', '333184', '333215', '333224', '360050', '402782', '402792', '402794', '402805', '402809']
nosy_count = 5.0
nosy_names = ['christian.heimes', 'pervlad', 'pedja', 'MDM-1', 'pukkandan']
pr_nums = []
priority = 'normal'
resolution = 'third party'
stage = 'resolved'
status = 'closed'
superseder = None
type = 'behavior'
url = 'https://bugs.python.org/issue35665'
versions = ['Python 3.9']
```

</p></details>

### Linked PRs
* gh-91740
* gh-122768
* gh-122769

### LINKED ISSUE #89475: "MUPCA Root" Certificates - treated as invalid and cause error, but are walid and necessary
BPO | [45312](https://bugs.python.org/issue45312)
--- | :---
Nosy | @pfmoore, @tjguk, @zware, @zooba, @MDM-79, @pukkandan
Dependencies | <li>bpo-35665: Function ssl.create_default_context raises exception on Windows 10  when called with  ssl.Purpose.SERVER_AUTH) attribute</li>
Superseder | <li>bpo-35665: Function ssl.create_default_context raises exception on Windows 10  when called with  ssl.Purpose.SERVER_AUTH) attribute</li>
Files | <li>[Untitled.png](https://bugs.python.org/file50312/Untitled.png "Uploaded as image/png at 2021-09-28.15:49:22 by @MDM-79"): MUPCA Root - Certificates in qestion</li><li>[135081521-83d466d9-c71d-465c-96a2-652c2549c461.png](https://bugs.python.org/file50313/135081521-83d466d9-c71d-465c-96a2-652c2549c461.png "Uploaded as image/png at 2021-09-28.15:58:15 by @MDM-79"): Error 1</li><li>[135087546-e6fd1b05-6858-4e0f-ad3f-857c614bb15b.png](https://bugs.python.org/file50314/135087546-e6fd1b05-6858-4e0f-ad3f-857c614bb15b.png "Uploaded as image/png at 2021-09-28.15:58:23 by @MDM-79"): Error 2</li>

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = None
closed_at = <Date 2021-09-28.16:11:10.264>
created_at = <Date 2021-09-28.15:49:22.455>
labels = ['expert-SSL', 'OS-windows', 'type-crash', '3.9']
title = '"MUPCA Root" Certificates - treated as invalid and cause error, but are walid and necessary'
updated_at = <Date 2021-09-28.21:57:36.256>
user = 'https://github.com/MDM-79'
```

bugs.python.org fields:
```python
activity = <Date 2021-09-28.21:57:36.256>
actor = 'steve.dower'
assignee = 'none'
closed = True
closed_date = <Date 2021-09-28.16:11:10.264>
closer = 'steve.dower'
components = ['Windows', 'SSL']
creation = <Date 2021-09-28.15:49:22.455>
creator = 'MDM-1'
dependencies = ['35665']
files = ['50312', '50313', '50314']
hgrepos = []
issue_num = 45312
keywords = []
message_count = 9.0
messages = ['402784', '402786', '402787', '402788', '402789', '402790', '402791', '402801', '402813']
nosy_count = 7.0
nosy_names = ['paul.moore', 'tim.golden', 'zach.ware', 'steve.dower', 'MDM-1', 'pukkandan', 'Pedjas']
pr_nums = []
priority = 'normal'
resolution = 'third party'
stage = 'resolved'
status = 'closed'
superseder = '35665'
type = 'crash'
url = 'https://bugs.python.org/issue45312'
versions = ['Python 3.9']
```

</p></details>

### PR DESCRIPTION
**Pull request by pukkandan**
Fixes #79846, fixes #89475

Currently, when loading certificates from the Windows certificate store, error in any one certificate causes `ssl.create_default_context()` to crash. This causes issues in systems that have certificates that are not quite to-spec. A primary culprit for this is ["MUPCA Root"](https://github.com/python/cpython/issues/79846#issuecomment-1093808502), which (despite being is technically invalid) is [essential for citizens of Serbia](https://github.com/python/cpython/issues/79846#issuecomment-1093808508)

See the conversations under the linked issues for more details

I believe it makes sense for `create_default_context` to ignore any invalid certificates in the system store. An existing comment in the related code seems to agree with me on this:
https://github.com/python/cpython/blob/84975146a7ce64f1d50dcec8311b7f7188a5c962/Lib/ssl.py#L772-L774

This issue can be solved by loading each certificate one by one and ignoring any `SSLError`s. I had [outlined the idea for this patch in the above-mentioned issue](https://github.com/python/cpython/issues/79846#issuecomment-1093808513), but never recieved any reply on whether this is acceptable. Hopefully, this PR receives better attention

* Issue: gh-79846

### CODE REVIEW

**Comment by serhiy-storchaka** [comment]
The issues was closed and I cannot reproduce the original problem.

**Comment by ideasman42** [comment]
This issue still seems to be a problem. 
While I don't use MS-Windows, multiple Blender users look to have run into this with Blender 4.2, using Python 3.11.
See: https://projects.blender.org/blender/blender/issues/124731#

File: Lib/ssl.py [hunk 1]
```diff
@@ -516,14 +516,17 @@ def _load_windows_store_certs(self, storename, purpose):
         certs = bytearray()
         try:
             for cert, encoding, trust in enum_certificates(storename):
-                # CA certs are never PKCS#7 encoded
-                if encoding == "x509_asn":
-                    if trust is True or purpose.oid in trust:
-                        certs.extend(cert)
+                try:
+                    self.load_verify_locations(cadata=cert)
```
Comment (serhiy-storchaka) [inline Lib/ssl.py]: This loads the certificate unconditionally, unlike the current code.

**Comment by zooba** [comment]
Is calling `load_verify_locations` multiple times okay? It doesn't replace the old loaded certs in any OpenSSL versions, right?

**Comment by zooba** [comment]
If it's tested, then that's fine. This change LGTM

**Pull request merged**